[api-documenter] - ExperimentalYamlDocumenter new filter feature addition

Vitalius1 opened the pull request:
1. Added accurate documentation to the config interface
2. Removed unnecessary config option `noDuplicateEntries`
3. Renamed some of the config options to give them an easier semantic meaning (open for any objections here 😄 )
4. Added new config option `nonEmptyCategoryNodeNames` which is very useful when needed to have as categories not only the nodes that have an empty array to be filled with API items but also already custom items like example items at the tome of the creation of the tree structure.
5. Refactored the categorization logic into a separate method and added the missing filter by name from previous iteration.
6. Modified the test package to verify the new config file (looks like it's working 😄 ).

natalieethell commented on apps/api-documenter/src/documenters/ExperimentalYamlDocumenter.ts:
```diff
@@ -118,11 +101,57 @@ export class ExperimentalYamlDocumenter extends YamlDocumenter {
   // This is a direct copy of a @docCategory inline tag finder in office-ui-fabric-react,
   // but is generic enough to be used for any inline tag
   private _findInlineTagByName(tagName: string, docComment: DocComment | undefined): DocInlineTag | undefined {
+    const tagNameToCheck: string = `@${tagName}`;
```
Good change. Are we still planning to take a dependency on api-docs and import `findInlineTagByName` from there?

Vitalius1 replied:
Not sure if we need to take a dependency on `api-docs` for this small utility helper function. `api-docs` is kinda a OUFR specific package and `api-documenter` should be generic. Perhaps the opposite way would make more sense. Have this extracted as an utility function maybe to `api-extractor-model` package and make it available to both places.

natalieethell replied:
Ah right, I remember talking about that. That makes sense to me!

Vitalius1 commented:
@octogonz friendly ping whenever you can find a few minutes. Thanks!

octogonz commented:
I'm very busy this week, but I'll see what I can do.

Vitalius1 commented:
@octogonz another friendly ping.
@iclanton, maybe you can take a look too? 🤔 Thanks!

iclanton commented on apps/api-documenter/.vscode/launch.json:
```diff
@@ -0,0 +1,23 @@
+{
```
Can you remove this file?

Vitalius1 replied:
Will do!

iclanton commented on apps/api-documenter/src/documenters/IConfigFile.ts:
```diff
@@ -9,37 +9,45 @@ import { IYamlTocFile } from '../yaml/IYamlTocFile';
 export interface IConfigTableOfContents {
```
>IConfigTableOfContents [](start = 17, length = 22)

Is there a schema file you need to update too?

Vitalius1 replied:
The only place I can think of it this file: https://github.com/microsoft/web-build-tools/blob/master/apps/api-documenter/src/schemas/api-documenter-template.json
@octogonz do I need to update this file with the changes I bring to `IConfigFile`? Where that template is being used?

iclanton commented on build-tests/api-documenter-test/src/DocClass1.ts:
```diff
@@ -204,3 +204,25 @@ export class DocClass1 extends DocBaseClass implements IDocInterface1, IDocInter
+  /**
+   * string property
+   */
+  string: string;
```
>string [](start = 2, length = 6)

Might be better to use member names that aren't also type names.

Vitalius1 replied:
fixed!

octogonz commented on build-tests/api-documenter-test/etc/api-documenter-test.api.md:
```diff
@@ -74,6 +74,16 @@ export interface IDocInterface4 {
 
+// @public
+export interface IDocInterface5 {
+    string: string;
```
Don't use a TypeScript keyword as a variable name.

Vitalius1 replied:
Fixed

octogonz commented on build-tests/api-documenter-test/etc/api-documenter-test.api.md:
```diff
@@ -74,6 +74,16 @@ export interface IDocInterface4 {
+
+// @public
+export interface IDocInterface6 {
+    number: number;
```
Don't use a TypeScript keyword as a variable name.

Vitalius1 replied:
Fixed!

octogonz commented on common/changes/@microsoft/api-documenter/vibraga-experimentalDocumenter_adjustments_2019-05-24-22-43.json:
```diff
@@ -0,0 +1,11 @@
+  "changes": [
+    {
+      "packageName": "@microsoft/api-documenter",
+      "comment": "Add new `nonEmptyCategoryNodeNames` config option to the ExperimentalYamlDocumenter. Refactor filter logic. Add documentation and modify the test package.",
```
This comment is describing implementation details.  Normally a change log should be written for an audience of people who use the tool (e.g. what feature did you add? what bug did you fix?), not people who work on the tool.

But since `ExperimentalYamlDocumenter` is experimental and not intended to be used by external people yet, I'd just say something simple like `Improved the experimental "api-documenter generate" command`.

Vitalius1 replied:
Fixed!

Vitalius1 commented:
@octogonz with the exception of that open question Ian asked about updating a schema I've addressed your feedback. Can you take another look and let me know if I need to update the schema template. Thanks!

octogonz commented:
> with the exception of that open question Ian asked about updating a schema I've addressed your feedback. Can you take another look and let me know if I need to update the schema template. Thanks!

@iclanton For now [api-documenter.schema.json](https://github.com/Vitalius1/web-build-tools/blob/vibraga/experimentalDocumenter_adjustments/apps/api-documenter/src/schemas/api-documenter.schema.json) is a placeholder that doesn't perform any actual validation.  The reason is that the file format is still just an experiment and hasn't been reviewed in depth, so it didn't want to waste time writing .schema.json for it.

There's a bunch of other settings that we should add to **api-documenter.json** unrelated to @Vitalius1's work here.  Basically I'm waiting for him to get to a point where he tells us "my prototype is now code complete, and it solves all the requirements for OUIFR."  Then we'll go back and review his approach, refactor the code, and generalize the config file so that it can also drive the other API Documenter scenarios.  Then we can graduate `api-documenter generate` to being an official command, replacing the `yaml` and `markdown` commands.

Vitalius1 commented:
@octogonz I am not authorized to merge this. I can't even add reviewers. Can you add me please to the contributors list as I expect to submit PRs in the future. Thanks!

octogonz commented:
> I am not authorized to merge this. I can't even add reviewers.

That's expected. We don't assign specific reviewers, and the repo maintainers merge PRs.

Vitalius1 commented:
So if I want Natalie or Kevin to review a PR I would just need to let them know about it?
